File: apps/backends/php/app/Services/CheckoutService.php
<?php

namespace App\Services;

use App\Models\Product;
use App\Models\Order;
use App\Models\OrderItem;
use Illuminate\Support\Str;
use Illuminate\Support\Facades\DB;

class CheckoutService
{
    private $taxService;
    private $fraudService;

    public function __construct(TaxService $taxService, FraudService $fraudService)
    {
        $this->taxService = $taxService;
        $this->fraudService = $fraudService;
    }

    public function processCheckout(string $customerId, array $items, string $state): array
    {
        // Validate items
        if (empty($items) || count($items) > 8) {
            throw new \Exception('Cart must have 1-8 items', 422);
        }

        return DB::transaction(function () use ($customerId, $items, $state) {
            $subtotal = 0;
            $orderItems = [];

            // Validate stock and calculate subtotal
            foreach ($items as $item) {
                $product = Product::whereRaw('id::text = ?', [$item['product_id']])->first();
                
                if (!$product) {
                    throw new \Exception('Product not found', 404);
                }

                if ($product->stock < $item['quantity']) {
                    throw new \Exception('Insufficient stock', 422);
                }

                $subtotal += $product->price_cents * $item['quantity'];
                $orderItems[] = [
                    'product_id' => $item['product_id'],
                    'quantity' => $item['quantity'],
                    'price_cents' => $product->price_cents,
                ];
            }

            // Calculate tax
            $taxResp = $this->taxService->calculateTax($subtotal, $state);
            $taxCents = $taxResp['tax_cents'] ?? 0;

            // Calculate fraud score
            $fraudScore = $this->fraudService->calculateScore($subtotal, count($items));

            // Create order
            $orderId = Str::uuid()->toString();
            $total = $subtotal + $taxCents;

            Order::create([
                'id' => $orderId,
                'customer_id' => $customerId,
                'total_cents' => $total,
                'tax_cents' => $taxCents,
                'created_at' => now(),
            ]);

            // Create order items
            foreach ($orderItems as $item) {
                OrderItem::create([
                    'id' => Str::uuid()->toString(),
                    'order_id' => $orderId,
                    'product_id' => $item['product_id'],
                    'quantity' => $item['quantity'],
                    'price_cents' => $item['price_cents'],
                    'created_at' => now(),
                ]);
            }

            return [
                'order_id' => $orderId,
                'total_cents' => $total,
                'tax_cents' => $taxCents,
                'fraud_score' => $fraudScore,
            ];
        }, attempts: 3, timeout: 5);
    }
}

File: apps/backends/rails/app/services/checkout_service.rb
require 'securerandom'

class CheckoutService
  UUID_RE = /\A[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\z/i

  def initialize(tax_client)
    @tax_client = tax_client
  end

  def call(params)
    cart           = params[:cart] || []
    customer_id    = params[:customer_id].to_s
    shipping       = params[:shipping_address] || {}

    raise ArgumentError, 'Cart must have 1-8 items'   if cart.empty? || cart.size > 8
    raise ArgumentError, 'Invalid customer ID'         unless UUID_RE.match?(customer_id)

    subtotal   = 0
    line_items = []

    cart.each do |item|
      product_id = item[:product_id].to_s
      quantity   = item[:quantity].to_i

      raise ArgumentError, 'Invalid product ID'          unless UUID_RE.match?(product_id)
      raise ArgumentError, 'Quantity must be 1-8'        unless (1..8).cover?(quantity)

      product = Product.find_by(id: product_id)
      raise StandardError, 'Product not found'           if product.nil?
      raise StandardError, 'Insufficient stock'          if product.stock < quantity

      subtotal += product.price_cents * quantity
      line_items << { product: product, quantity: quantity, price_cents: product.price_cents }
    end

    tax_resp    = @tax_client.calculate(subtotal_cents: subtotal, state: shipping[:country])
    tax_cents   = tax_resp[:tax_cents].to_i
    total_cents = subtotal + tax_cents
    fraud_score = (subtotal / 100) + (line_items.size * 10)
    order_id    = SecureRandom.uuid

    ActiveRecord::Base.transaction do
      Order.create!(
        id:          order_id,
        customer_id: customer_id,
        total_cents: total_cents,
        tax_cents:   tax_cents,
        created_at:  Time.now.utc
      )

      line_items.each do |li|
        OrderItem.create!(
          id:          SecureRandom.uuid,
          order_id:    order_id,
          product_id:  li[:product].id,
          quantity:    li[:quantity],
          price_cents: li[:price_cents],
          created_at:  Time.now.utc
        )

        updated = Product.where(id: li[:product].id)
                         .where('stock >= ?', li[:quantity])
                         .update_all('stock = stock - ?', li[:quantity])
        raise StandardError, 'Insufficient stock' if updated == 0
      end
    end

    {
      order_id:    order_id,
      total_cents: total_cents,
      tax_cents:   tax_cents,
      fraud_score: fraud_score,
      items:       line_items.map do |li|
        {
          product_id:   li[:product].id,
          product_name: li[:product].name,
          quantity:     li[:quantity],
          price_cents:  li[:price_cents]
        }
      end,
      created_at: Time.now.utc.iso8601
    }
  end
end
